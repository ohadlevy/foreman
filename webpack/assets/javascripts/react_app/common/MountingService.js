import React from 'react';
import StatisticsChartsList from '../components/charts/StatisticsChartsList';
import FlashNotifications from '../components/notifications/FlashNotifications';

import ReactDOM from 'react-dom';

export function mount(component, selector, data) {

  const components = {
    StatisticsChartsList: {
      type: StatisticsChartsList,
      markup: <StatisticsChartsList data={data}/>
    },
    FlashNotifications: {
     type: FlashNotifications,
     markup: <FlashNotifications flash={data}/>
   }
  };

  const reactNode = document.querySelector(selector);

  if (reactNode) {
    ReactDOM.render(components[component].markup, reactNode);
  } else {
    const componentName = components[component].type.name;

    // eslint-disable-next-line no-console
    console.log(`Cannot find \'${selector}\' element for mounting the \'${componentName}\'`);
  }
}
