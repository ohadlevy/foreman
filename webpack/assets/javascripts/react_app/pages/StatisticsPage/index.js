import React from 'react';
import PropTypes from 'prop-types';
import PageLayout from '../common/PageLayout/PageLayout';
import StatisticsChartsList from '../../components/statistics/StatisticsChartsList';

const Statistics = () => (
  <PageLayout header={__('Statistics')} searchable={false}>
    <StatisticsChartsList
      data={[
        {
          id: 'operatingsystem',
          title: 'OS Distribution',
          url: '/statistics/operatingsystem',
          search: '/hosts?search=os_title=~VAL~',
        },
        {
          id: 'architecture',
          title: 'Architecture Distribution',
          url: '/statistics/architecture',
          search: '/hosts?search=facts.architecture=~VAL~',
        },
        {
          id: 'environment',
          title: 'Environment Distribution',
          url: '/statistics/environment',
          search: '/hosts?search=environment=~VAL~',
        },
        {
          id: 'hostgroup',
          title: 'Host Group Distribution',
          url: '/statistics/hostgroup',
          search: '/hosts?search=hostgroup_title=~VAL~',
        },
        {
          id: 'compute_resource',
          title: 'Compute Resource Distribution',
          url: '/statistics/compute_resource',
          search: '/hosts?search=compute_resource=~VAL~',
        },
        {
          id: 'processorcount',
          title: 'Number of CPUs',
          url: '/statistics/processorcount',
          search: '/hosts?search=facts.processorcount=~VAL1~',
        },
        {
          id: 'manufacturer',
          title: 'Hardware',
          url: '/statistics/manufacturer',
          search: '/hosts?search=facts.manufacturer~~VAL~',
        },
        {
          id: 'memory',
          title: 'Average Memory Usage',
          url: '/statistics/memory',
          search: '/hosts?search=',
        },
        {
          id: 'swap',
          title: 'Average Swap Usage',
          url: '/statistics/swap',
          search: '/hosts?search=',
        },
        {
          id: 'puppetclass',
          title: 'Class Distribution',
          url: '/statistics/puppetclass',
          search: '/hosts?search=class=~VAL1~',
        },
        {
          id: 'location',
          title: 'Location Distribution',
          url: '/statistics/location',
          search: '/hosts?search=location=~VAL~',
        },
        {
          id: 'organization',
          title: 'Organization Distribution',
          url: '/statistics/organization',
          search: '/hosts?search=organization=~VAL~',
        },
      ]}
    />
  </PageLayout>
);

export default Statistics;
