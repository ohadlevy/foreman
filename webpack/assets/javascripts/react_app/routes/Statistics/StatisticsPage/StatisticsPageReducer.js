import Immutable from 'seamless-immutable';

import {
  STATISTICS_PAGE_META_RESOLVED,
  STATISTICS_PAGE_META_FAILED,
  STATISTICS_PAGE_HIDE_LOADING,
  STATISTICS_DATA_REQUEST,
  STATISTICS_DATA_SUCCESS,
  STATISTICS_DATA_FAILURE,
} from '../constants';

const initialState = Immutable({
  statisticsMeta: [],
  charts: Immutable({}),
  message: { type: 'empty', text: '' },
  isLoading: true,
  hasError: false,
});

export default (state = initialState, action) => {
  const { payload } = action;

  switch (action.type) {
    case STATISTICS_PAGE_META_RESOLVED:
      return state.set('statisticsMeta', payload);
    case STATISTICS_PAGE_HIDE_LOADING:
      return state.set('isLoading', false);
    case STATISTICS_PAGE_META_FAILED:
      return state.merge(payload);
    case STATISTICS_DATA_REQUEST:
      return state.setIn(['charts', payload.id], payload);
    case STATISTICS_DATA_SUCCESS:
      return state.setIn(['charts', payload.id], {
        ...state.charts[payload.id],
        data: payload.data,
      });
    case STATISTICS_DATA_FAILURE:
      return state.setIn(['charts', payload.item.id], {
        ...state.charts[payload.item.id],
        error: payload.error,
      });

    default:
      return state;
  }
};
