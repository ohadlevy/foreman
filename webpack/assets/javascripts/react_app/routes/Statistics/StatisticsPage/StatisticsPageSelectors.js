export const selectStatisticsPage = state => state.statisticsPage;
export const selectStatisticsCharts = state =>
  selectStatisticsPage(state).charts;
export const selectStatisticsMetadata = state =>
  selectStatisticsPage(state).statisticsMeta;
export const selectStatisticsIsLoading = state =>
  selectStatisticsPage(state).isLoading;
export const selectStatisticsMessage = state =>
  selectStatisticsPage(state).message;
export const selectStatisticsHasError = state =>
  selectStatisticsPage(state).hasError;
export const selectStatisticsHasMetadata = state =>
  selectStatisticsMetadata(state).length > 0;
