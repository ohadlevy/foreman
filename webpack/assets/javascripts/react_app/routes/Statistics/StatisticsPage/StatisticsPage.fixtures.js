import { statisticsData } from '../../../components/statistics/StatisticsChartsList.fixtures';
import { noop } from '../../../common/helpers';

export const statisticsProps = {
  statisticsMeta: statisticsData,
  charts: [
    {
      data: [['centOS 7.1', 6]],
      id: 'operatingsystem',
      search: '/hosts?search=os_title=~VAL~',
      title: 'OS Distribution',
      url: 'statistics/operatingsystem',
    },
  ],
  isLoading: false,
  hasData: true,
  hasError: false,
  message: {},
  getStatisticsMeta: noop,
  getStatisticsData: noop,
};

export const successRequestData = [
  {
    id: 'operatingsystem',
    title: 'OS Distribution',
    url: 'statistics/operatingsystem',
    search: '/hosts?search=os_title=~VAL~',
  },
  {
    id: 'architecture',
    title: 'Architecture Distribution',
    url: 'statistics/architecture',
    search: '/hosts?search=facts.architecture=~VAL~',
  },
];

export const request = {
  id: 'operatingsystem',
  title: 'OS Distribution',
  url: 'statistics/operatingsystem',
  search: '/hosts?search=os_title=~VAL~',
};

export const response = {
  id: 'operatingsystem',
  data: [['RedHat 3', 2]],
};
