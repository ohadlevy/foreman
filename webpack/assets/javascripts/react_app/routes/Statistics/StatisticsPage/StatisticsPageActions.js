import API from '../../../API';

import {
  STATISTICS_PAGE_META_RESOLVED,
  STATISTICS_PAGE_META_FAILED,
  STATISTICS_PAGE_HIDE_LOADING,
  STATISTICS_PAGE_URL,
  STATISTICS_DATA_REQUEST,
  STATISTICS_DATA_SUCCESS,
  STATISTICS_DATA_FAILURE,
} from '../constants';
import { ajaxRequestAction } from '../../../redux/actions/common';

export const getStatisticsMeta = (
  url = STATISTICS_PAGE_URL
) => async dispatch => {
  const onFetchSuccess = ({ data }) => {
    dispatch(hideLoading());
    dispatch({
      type: STATISTICS_PAGE_META_RESOLVED,
      payload: data,
    });
  };

  const onFetchError = ({ message }) => {
    dispatch(hideLoading());
    dispatch({
      type: STATISTICS_PAGE_META_FAILED,
      payload: {
        hasError: true,
        message: {
          type: 'error',
          text: message,
        },
      },
    });
  };
  try {
    const response = await API.get(url);
    return onFetchSuccess(response);
  } catch (error) {
    return onFetchError(error);
  }
};

export const getStatisticsData = charts => dispatch =>
  Promise.all(
    charts.map(chart =>
      ajaxRequestAction({
        dispatch,
        requestAction: STATISTICS_DATA_REQUEST,
        successAction: STATISTICS_DATA_SUCCESS,
        failedAction: STATISTICS_DATA_FAILURE,
        url: chart.url,
        item: chart,
      })
    )
  );

const hideLoading = () => ({
  type: STATISTICS_PAGE_HIDE_LOADING,
});
