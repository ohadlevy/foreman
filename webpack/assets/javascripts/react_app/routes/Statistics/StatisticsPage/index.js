import { compose, bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { callOnMount } from '../../../common/HOC';

import * as actions from './StatisticsPageActions';
import reducer from './StatisticsPageReducer';
import {
  selectStatisticsMetadata,
  selectStatisticsMessage,
  selectStatisticsIsLoading,
  selectStatisticsHasMetadata,
  selectStatisticsHasError,
  selectStatisticsCharts,
} from './StatisticsPageSelectors';

import StatisticsPage from './StatisticsPage';

// map state to props
const mapStateToProps = state => ({
  statisticsMeta: selectStatisticsMetadata(state),
  charts: selectStatisticsCharts(state),
  isLoading: selectStatisticsIsLoading(state),
  message: selectStatisticsMessage(state),
  hasData: selectStatisticsHasMetadata(state),
  hasError: selectStatisticsHasError(state),
});

// map action dispatchers to props
const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);

// export reducers
export const reducers = { statisticsPage: reducer };

// export connected component
export default compose(
  connect(
    mapStateToProps,
    mapDispatchToProps
  ),
  callOnMount(({ getStatisticsMeta }) => getStatisticsMeta())
)(StatisticsPage);
