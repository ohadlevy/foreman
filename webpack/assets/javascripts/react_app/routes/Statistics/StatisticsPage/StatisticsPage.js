import React from 'react';
import PropTypes from 'prop-types';
import PageLayout from '../../../pages/common/PageLayout/PageLayout';
import Statistics from './Statistics/Statistics';

const StatisticsPage = ({ statisticsMeta, ...props }) => (
  <PageLayout header={__('Statistics')} searchable={false}>
    <Statistics statisticsMeta={statisticsMeta} {...props} />
  </PageLayout>
);

StatisticsPage.propTypes = {
  statisticsMeta: PropTypes.array.isRequired,
};

export default StatisticsPage;
