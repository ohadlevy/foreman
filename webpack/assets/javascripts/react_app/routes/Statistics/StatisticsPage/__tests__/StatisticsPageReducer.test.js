import { testReducerSnapshotWithFixtures } from '../../../../common/testHelpers';
import { request, response } from '../StatisticsPage.fixtures';

import {
  STATISTICS_PAGE_META_RESOLVED,
  STATISTICS_PAGE_META_FAILED,
  STATISTICS_PAGE_HIDE_LOADING,
  STATISTICS_DATA_REQUEST,
  STATISTICS_DATA_SUCCESS,
  STATISTICS_DATA_FAILURE,
} from '../../constants';
import reducer from '../StatisticsPageReducer';

const fixtures = {
  'should return the initial state': {},
  'should handle STATISTICS_PAGE_FETCH_META': {
    action: {
      type: STATISTICS_PAGE_META_RESOLVED,
      payload: [],
    },
  },
  'should handle STATISTICS_PAGE_HIDE_LOADING': {
    action: {
      type: STATISTICS_PAGE_HIDE_LOADING,
    },
  },
  'should handle STATISTICS_PAGE_SHOW_MESSAGE': {
    action: {
      type: STATISTICS_PAGE_META_FAILED,
      payload: {
        hasError: true,
        message: { type: 'error', text: 'some-error' },
      },
    },
  },
  'should handle STATISTICS_DATA_REQUEST': {
    action: {
      type: STATISTICS_DATA_REQUEST,
      payload: request,
    },
  },
  'should handle STATISTICS_DATA_SUCCESS': {
    action: {
      type: STATISTICS_DATA_SUCCESS,
      payload: response,
    },
  },
  'should handle STATISTICS_DATA_FAILURE': {
    action: {
      type: STATISTICS_DATA_FAILURE,
      payload: { error: 'some-error', item: request },
    },
  },
};

describe('StatisticsPage reducer', () =>
  testReducerSnapshotWithFixtures(reducer, fixtures));
