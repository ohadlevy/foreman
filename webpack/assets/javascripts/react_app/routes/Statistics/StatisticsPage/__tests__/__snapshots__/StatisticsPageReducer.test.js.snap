// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatisticsPage reducer should handle STATISTICS_DATA_FAILURE 1`] = `
Object {
  "charts": Object {
    "operatingsystem": Object {
      "error": "some-error",
    },
  },
  "hasError": false,
  "isLoading": true,
  "message": Object {
    "text": "",
    "type": "empty",
  },
  "statisticsMeta": Array [],
}
`;

exports[`StatisticsPage reducer should handle STATISTICS_DATA_REQUEST 1`] = `
Object {
  "charts": Object {
    "operatingsystem": Object {
      "id": "operatingsystem",
      "search": "/hosts?search=os_title=~VAL~",
      "title": "OS Distribution",
      "url": "statistics/operatingsystem",
    },
  },
  "hasError": false,
  "isLoading": true,
  "message": Object {
    "text": "",
    "type": "empty",
  },
  "statisticsMeta": Array [],
}
`;

exports[`StatisticsPage reducer should handle STATISTICS_DATA_SUCCESS 1`] = `
Object {
  "charts": Object {
    "operatingsystem": Object {
      "data": Array [
        Array [
          "RedHat 3",
          2,
        ],
      ],
    },
  },
  "hasError": false,
  "isLoading": true,
  "message": Object {
    "text": "",
    "type": "empty",
  },
  "statisticsMeta": Array [],
}
`;

exports[`StatisticsPage reducer should handle STATISTICS_PAGE_FETCH_META 1`] = `
Object {
  "charts": Object {},
  "hasError": false,
  "isLoading": true,
  "message": Object {
    "text": "",
    "type": "empty",
  },
  "statisticsMeta": Array [],
}
`;

exports[`StatisticsPage reducer should handle STATISTICS_PAGE_HIDE_LOADING 1`] = `
Object {
  "charts": Object {},
  "hasError": false,
  "isLoading": false,
  "message": Object {
    "text": "",
    "type": "empty",
  },
  "statisticsMeta": Array [],
}
`;

exports[`StatisticsPage reducer should handle STATISTICS_PAGE_SHOW_MESSAGE 1`] = `
Object {
  "charts": Object {},
  "hasError": true,
  "isLoading": true,
  "message": Object {
    "text": "some-error",
    "type": "error",
  },
  "statisticsMeta": Array [],
}
`;

exports[`StatisticsPage reducer should return the initial state 1`] = `
Object {
  "charts": Object {},
  "hasError": false,
  "isLoading": true,
  "message": Object {
    "text": "",
    "type": "empty",
  },
  "statisticsMeta": Array [],
}
`;
